### PR TITLE
Adding comments to a lot of hs_core root files.

### PR DESCRIPTION
**Pull request by aphelionz**
Left to do

- [x] hs_core/models.py
- [x] hs_core/search_indexes.py
- [x] hs_core/tasks.py
- [x] hs_core/testing.py

### CODE REVIEW

File: hs_core/forms.py [hunk 1]
```diff
@@ -980,8 +1133,8 @@ def clean(self):
 
 
 def get_crispy_form_fields(field_names, file_type=False):
-    """
-    This creates a list of objects of type Field
+    """Return a list of objects of type Field.
```
Comment (hyi) [inline hs_core/forms.py]: @aphelionz missing the closing pair of ```"""```

File: hs_core/forms.py [hunk 2]
```diff
@@ -980,8 +1133,8 @@ def clean(self):
...
+
```
Comment (hyi) [inline hs_core/forms.py]: @aphelionz Retract my previous comment - perhaps making ```"""``` a separate line is clearer.

Comment (aphelionz) [inline hs_core/forms.py]: Believe it or not `"""This format is flake8 standard`

File: hs_core/receivers.py [hunk 3]
```diff
@@ -1,6 +1,8 @@
+"""Signal receivers for the hs_core app.
 
-# Note: this module has been imported in the models.py in order to receive signals
-# se the end of the models.py for the import of this module
+Note: this module has been imported in the models.py in order to receive signals
+see the end of the models.py for the import of this module
+"""
```
Comment (hyi) [inline hs_core/receivers.py]: @aphelionz I believe this comment does not apply any more. The import of this module has been moved to HSCoreAppConfig class instead.

File: hs_core/forms.py (hunk 2 above)
Comment (hyi) [inline hs_core/forms.py]: @aphelionz I see you have used this commenting format consistently, so it is your call which format works better and you can ignore my comment above.

File: hs_core/forms.py [hunk 4]
```diff
@@ -190,22 +225,28 @@ def __init__(self, allow_edit=True, res_short_id=None, element_id=None, *args, *
...
+        """Describe meta properties of PartyForm."""
+
         model = Creator
         fields = PartyForm.Meta.fields
         fields.append("order")
         labels = PartyForm.Meta.labels
 
 
 class PartyValidationForm(forms.Form):
+    """Render validated form for Party models."""
```
Comment (pkdash) [inline hs_core/forms.py]: This form (and other validation forms) is used only for validating user posted data for creating/updating party models. This form is not used for rendering.

File: hs_core/forms.py [hunk 5]
```diff
@@ -500,7 +601,10 @@ def __init__(self, *args, **kwargs):
 
 
 class FundingAgencyForm(ModelForm):
+    """Render Identifier model form with appropriate attributes."""
```
Comment (pkdash) [inline hs_core/forms.py]: ```Identifier``` -> ```FundingAgency```

File: hs_core/forms.py [hunk 6]
```diff
@@ -500,7 +601,10 @@ def __init__(self, *args, **kwargs):
...
+
     def __init__(self, allow_edit=True, res_short_id=None, *args, **kwargs):
+        """Render Identifier model form with appropriate attributes."""
```
Comment (pkdash) [inline hs_core/forms.py]: As above

File: hs_core/forms.py [hunk 7]
```diff
@@ -621,12 +741,15 @@ def __init__(self, allow_edit=True, res_short_id=None, element_id=None, element_
 
 
 class SubjectsForm(forms.Form):
+    """Render Identifier model form with appropriate attributes."""
```
Comment (pkdash) [inline hs_core/forms.py]: ```Identifier``` -> ```Subject```

File: hs_core/forms.py [hunk 8]
```diff
@@ -621,12 +741,15 @@ def __init__(self, allow_edit=True, res_short_id=None, element_id=None, element_
...
+
     value = forms.CharField(max_length=500,
                             label='',
                             widget=forms.TextInput(attrs={'placeholder': 'Keywords'}),
                             help_text='Enter each keyword separated by a comma.')
 
     def __init__(self, allow_edit=True, res_short_id=None, element_id=None, *args, **kwargs):
+        """Render Identifier model form with appropriate attributes."""
```
Comment (pkdash) [inline hs_core/forms.py]: As above

File: hs_core/forms.py [hunk 9]
```diff
@@ -659,7 +786,10 @@ def __init__(self, allow_edit=True, res_short_id=None, element_id=None, element_
 
 
 class AbstractForm(ModelForm):
+    """Render Abstract model form with appropriate attributes."""
+
     def __init__(self, allow_edit=True, res_short_id=None, element_id=None, *args, **kwargs):
+        """Render Identifier model form with appropriate attributes."""
```
Comment (pkdash) [inline hs_core/forms.py]: ```Identifier``` -> ```Abstract```

**Comment by pkdash** [comment]
@aphelionz I have now reviewed and left few comments. I assume you plan to add similar docstrings to rest of the files in hs_core as part of this PR.

**Comment by aphelionz** [comment]
@pkdash @hyi Comments were addressed and the rest of the .py files in the root were commented. Thank you!

File: hs_core/models.py [hunk 10]
```diff
@@ -77,46 +81,58 @@ def validate_user_url(value):
...
+
         abstract = True
 
     @property
     def permissions_store(self):
+        """Use PERMISSIONS_DB constant. Unsure what 's' is here."""
         return s.PERMISSIONS_DB
 
     def can_add(self, request):
+        """Pass through can_change to determine if user can_add resources or resource files."""
```
Comment (pkdash) [inline hs_core/models.py]: Any user can add resources in hydroshare. So may be better to document as: 

Pass through can_change to determine if user can make changes to a resource.

File: hs_core/models.py [hunk 11]
```diff
@@ -77,46 +81,58 @@ def validate_user_url(value):
...
         return self.can_change(request)
 
     def can_delete(self, request):
+        """Use utils.authorize method to determine if user can delete."""
```
Comment (pkdash) [inline hs_core/models.py]: ... if user can delete. - > ... if user can delete a resource.

File: hs_core/models.py [hunk 12]
```diff
@@ -77,46 +81,58 @@ def validate_user_url(value):
...
         # have to do import locally to avoid circular import
         from hs_core.views.utils import authorize, ACTION_TO_AUTHORIZE
         return authorize(request, self.short_id,
                          needed_permission=ACTION_TO_AUTHORIZE.DELETE_RESOURCE,
                          raises_exception=False)[1]
 
     def can_change(self, request):
+        """Use utils.authorize method to determine if user can change."""
```
Comment (pkdash) [inline hs_core/models.py]: ... if user can change -> ... if user can change/edit a resource.

File: hs_core/models.py [hunk 13]
```diff
@@ -77,46 +81,58 @@ def validate_user_url(value):
...
         # have to do import locally to avoid circular import
         from hs_core.views.utils import authorize, ACTION_TO_AUTHORIZE
         return authorize(request, self.short_id,
                          needed_permission=ACTION_TO_AUTHORIZE.EDIT_RESOURCE,
                          raises_exception=False)[1]
 
     def can_view(self, request):
+        """Use utils.authorize method to determine if user can view."""
```
Comment (pkdash) [inline hs_core/models.py]: ... if user can view -> ... if user can view a resource.

File: hs_core/models.py [hunk 14]
```diff
@@ -77,46 +81,58 @@ def validate_user_url(value):
...
                          raises_exception=False)[1]
 
 
-# this should be used as the page processor for anything with pagepermissionsmixin
-# page_processor_for(MyPage)(ga_resources.views.page_permissions_page_processor)
 def page_permissions_page_processor(request, page):
+    """Return a dict describing permissions for current user.
+
+    This should be used as the page processor for anything with pagepermissionsmixin
+    page_processor_for(MyPage)(ga_resources.views.page_permissions_page_processor)
```
Comment (pkdash) [inline hs_core/models.py]: ga_resources module is no longer part of Hydroshare.

File: hs_core/models.py [hunk 15]
```diff
@@ -948,17 +1057,18 @@ class Meta:
 
     @property
     def value(self):
+        """Return json representation of coverage values."""
         return json.loads(self._value)
 
     @classmethod
     def create(cls, **kwargs):
-        """
+        """Define custom create method for Language model.
```
Comment (pkdash) [inline hs_core/models.py]: Language -> Covergae

File: hs_core/models.py [hunk 16]
```diff
@@ -1022,10 +1132,10 @@ def create(cls, **kwargs):
 
     @classmethod
     def update(cls, element_id, **kwargs):
-        """
+        """Define custom create method for Language model.
```
Comment (pkdash) [inline hs_core/models.py]: Language -> Coverage

File: hs_core/models.py [hunk 17]
```diff
@@ -1069,9 +1179,11 @@ def update(cls, element_id, **kwargs):
 
     @classmethod
     def remove(cls, element_id):
+        """Define custom remove method for Language model."""
```
Comment (pkdash) [inline hs_core/models.py]: Language -> Coverage

File: hs_core/models.py [hunk 18]
```diff
@@ -2835,66 +2994,74 @@ def remove_folder(cls, resource, folder, user):
...
+        """Return filename of resource file."""
         # from .hydroshare.utils import get_resource_file_name_and_extension
         # return get_resource_file_name_and_extension(self)[1]
         return os.path.basename(self.storage_path)
 
     @property
     def can_set_file_type(self):
-        # currently user can set file type only for files with extension
-        # tif, zip and nc.
+        """Allow user to set file type on tif, zip and nc."""
```
Comment (pkdash) [inline hs_core/models.py]: """Check if file type can be set for this resource file instance."""

File: hs_core/models.py [hunk 19]
```diff
@@ -2930,26 +3096,37 @@ def irods_url(self):
...
 class PublicResourceManager(models.Manager):
+    """Extend Django model Manager to allow for public resource access."""
+
     def get_queryset(self):
+        """Extend Django model Manager to allow for public resource access."""
         return super(PublicResourceManager, self).get_queryset().filter(raccess__public=True)
 
 
 class DiscoverableResourceManager(models.Manager):
+    """Extend Django model Manager to filter for public and discoverable resources."""
```
Comment (pkdash) [inline hs_core/models.py]: ... public and discoverable resources. -> ...public **or** discoverable resources.

File: hs_core/models.py [hunk 20]
```diff
@@ -2979,37 +3157,44 @@ class BaseResource(Page, AbstractResource):
...
     def can_delete(self, request):
+        """Pass through to abstract resource can_delete function."""
         return AbstractResource.can_delete(self, request)
 
     def can_view(self, request):
+        """Pass through to abstract resource can_view function."""
         return AbstractResource.can_view(self, request)
 
     def get_irods_storage(self):
+        """Return either IrodsStorage and FedStorage."""
```
Comment (pkdash) [inline hs_core/models.py]: ... IrodsStorage and FedStorage. -> IrodsStorage or FedStorage.

File: hs_core/models.py [hunk 21]
```diff
@@ -3760,21 +3967,25 @@ def _create_person_element(self, etree, parent_element, person):
             hsterms_link_type.set('{%s}resource' % self.NAMESPACES['rdf'], link.url)
 
     def create_element(self, element_model_name, **kwargs):
+        """Create generic metadata element."""
```
Comment (pkdash) [inline hs_core/models.py]: """Create any supported metadata element."""

File: hs_core/page_processors.py [hunk 22]
```diff
@@ -19,15 +21,15 @@
 
 @processor_for(GenericResource)
 def landing_page(request, page):
+    """Return landing page context."""
```
Comment (pkdash) [inline hs_core/page_processors.py]: """Return resource landing page context."""

File: hs_core/testing.py [hunk 23]
```diff
@@ -33,22 +38,26 @@ def setUp(self):
...
     def tearDown(self):
+        """Stop iRODS patchers."""
         if settings.IRODS_HOST != 'data.local.org':
             for patcher in self.irods_patchers:
                 patcher.stop()
         super(MockIRODSTestCaseMixin, self).tearDown()
 
 
 class TestCaseCommonUtilities(object):
+    """Enable commun utilities for iRODS testing."""
```
Comment (pkdash) [inline hs_core/testing.py]: typo

File: hs_core/testing.py [hunk 24]
```diff
@@ -256,7 +266,8 @@ def resource_file_oprs(self):
         self.assertEqual(updated_res_file_names[0], 'new_' + file_name_list[2])
 
     def raster_metadata_extraction(self):
-        """
+        """Test Metadata extraction.
```
Comment (pkdash) [inline hs_core/testing.py]: Test raster metadata extraction

File: hs_core/search_indexes.py [hunk 25]
```diff
@@ -288,51 +317,61 @@ def prepare_coverage_end_date(self, obj):
             return 'none'
 
     def prepare_formats(self, obj):
+        """Return metadata formats if metadat exists, otherwise return empty array."""
```
Comment (pkdash) [inline hs_core/search_indexes.py]: typo

File: hs_core/search_indexes.py [hunk 26]
```diff
@@ -288,51 +317,61 @@ def prepare_coverage_end_date(self, obj):
...
     def prepare_comments(self, obj):
+        """Return list of all comments on resource."""
         return [comment.comment for comment in obj.comments.all()]
 
     def prepare_comments_count(self, obj):
+        """Return count of resource comments."""
         return obj.comments_count
 
     def prepare_owners_logins(self, obj):
+        """Return list of usernames that have access to resource."""
```
Comment (pkdash) [inline hs_core/search_indexes.py]: ... that have ownership access to resource.

**Comment by pkdash** [comment]
@aphelionz I have left few comments for your consideration.

File: hs_core/models.py (hunk 18 above)
Comment (aphelionz) [inline hs_core/models.py]: This was resolved, not sure why it's not showing up here

**Comment by pkdash** [comment]
+1

**Pull request merged**